src/hmacvalidator/hmacvalidator.go

```diff
@@ -0,0 +1,80 @@
+package hmacvalidator
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/adyen/adyen-go-api-library/src/notification"
+)
+
+func CalculateHmac(data interface{}, secret string) (string, error) {
+	switch val := data.(type) {
+	case string:
+		return encode(val, secret)
+	default:
+		src := GetDataToSign(data)
+		return encode(src, secret)
+	}
+}
+
+func ValidateHmac(notificationRequestItem notification.NotificationRequestItem, key string) bool {
+	expectedSign, err := CalculateHmac(notificationRequestItem, key)
+	if err != nil {
+		return false
+	}
+	merchantSign := (*notificationRequestItem.AdditionalData)["HmacSignature"]
+	return expectedSign == merchantSign
+}
+
+func GetDataToSign(notificationRequestItem interface{}) string {
+	switch item := notificationRequestItem.(type) {
+	case notification.NotificationRequestItem:
+		signedDataList := []string{
+			item.PspReference,
+			item.OriginalReference,
+			item.MerchantAccountCode,
+			item.MerchantReference,
+			strconv.Itoa(int(item.Amount.Value)),
+			item.Amount.Currency,
+			item.EventCode,
+			item.Success,
+		}
+		return strings.Join(signedDataList, ":")
+	case map[string]string:
+		keys := make([]string, 0)
+		values := make([]string, 0)
+
+		for k, v := range item {
+			keys = append(keys, replacer(k))
+			values = append(values, replacer(v))
+		}
+
+		return strings.Join(keys, ":") + ":" + strings.Join(values, ":")
```

Comment:
oh most important thing, maps in go are not sorted and you cannot sort them, means you cannot rely on their order for anything, so this won't work. What kind of order is expected in output, maybe you can sort the key and value array